op-challenger/game/keccak/matrix/matrix_test.go

```diff
@@ -220,6 +195,126 @@ func TestMatrix_AbsorbNextLeaf(t *testing.T) {
 	}
 }
 
+func TestVerifyPreimage(t *testing.T) {
+	preimage := testutils.RandomData(rand.New(rand.NewSource(2323)), 1024)
+	validCommitments := func() []common.Hash {
+		s := NewStateMatrix()
+		valid, err := s.AbsorbUpTo(bytes.NewReader(preimage), 1000*types.BlockSize)
+		require.ErrorIs(t, err, io.EOF, "Should read all preimage data")
+		return valid.Commitments
+	}
+	leafData := func(idx int) [types.BlockSize]byte {
+		var out [types.BlockSize]byte
+		copy(out[:], preimage[idx*types.BlockSize:(idx+1)*types.BlockSize])
+		return out
+	}
+	challengeLeaf := func(commitments []common.Hash, invalidIdx int) *types.Challenge {
+		invalidLeafStart := invalidIdx * types.BlockSize
+		s := NewStateMatrix()
+		_, err := s.AbsorbUpTo(bytes.NewReader(preimage), invalidLeafStart)
+		require.NoError(t, err)
+
+		return &types.Challenge{
+			StateMatrix: s.PackState(),
+			Prestate: types.Leaf{
+				Input:           leafData(invalidIdx - 1),
+				Index:           big.NewInt(int64(invalidIdx - 1)),
+				StateCommitment: commitments[invalidIdx-1],
+			},
+			//PrestateProof: nil,
+			Poststate: types.Leaf{
+				Input:           leafData(invalidIdx),
+				Index:           big.NewInt(int64(invalidIdx)),
+				StateCommitment: commitments[invalidIdx],
+			},
+			//PoststateProof: nil,
+		}
+	}
+
+	type testInputs struct {
+		name        string
+		commitments func() []common.Hash
+		expected    *types.Challenge
+	}
+
+	tests := []testInputs{
+		{
+			name:        "Valid",
+			commitments: validCommitments,
+			expected:    nil,
+		},
+		{
+			name: "IncorrectFirstLeaf",
+			commitments: func() []common.Hash {
+				commitments := validCommitments()
+				commitments[0] = common.Hash{0xaa}
+				return commitments
+			},
+			expected: &types.Challenge{
+				StateMatrix: NewStateMatrix().PackState(),
+				Prestate:    types.Leaf{},
+				//PrestateProof: nil,
+				Poststate: types.Leaf{
+					Input:           leafData(0),
+					Index:           big.NewInt(int64(0)),
+					StateCommitment: common.Hash{0xaa},
+				},
+				//PoststateProof: nil,
+			},
+		},
+	}
+
+	for i := 1; i < len(preimage)/types.BlockSize; i++ {
+		commitments := validCommitments()
+		commitments[i] = common.Hash{0xaa}
+		tests = append(tests, testInputs{
+			name: fmt.Sprintf("Incorrect-%v", i),
+			commitments: func() []common.Hash {
+				return commitments
+			},
+			expected: challengeLeaf(commitments, i),
+		})
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			challenge, err := VerifyPreimage(bytes.NewReader(preimage), test.commitments())
+			require.NoError(t, err)
+			require.Equal(t, test.expected, challenge)
+		})
+	}
+}
+
+func TestVerifyPreimage_DataMultipleOfBlockSize(t *testing.T) {
+	preimage := testutils.RandomData(rand.New(rand.NewSource(2323)), 5*types.BlockSize)
+	s := NewStateMatrix()
+	valid, err := s.AbsorbUpTo(bytes.NewReader(preimage), 1000*types.BlockSize)
```

Comment:
If this isn't for line length formatting, can also inline the `NewStateMatrix()` here. Feel free to dismiss if it is